File: Python/textbook_track/chapter03/ba3k.py
# Generate Contigs from a Collection of Reads
import itertools as it
import functools as ft
import sys

from collections import defaultdict


class Graph:
    def __init__(self, k_mers):
        self._adjacency_list = build_de_bruijn_graph(k_mers)
        self._in_degrees, self._out_degrees = self._calc_node_degrees()

    @ft.cached_property
    def nodes(self):
        nodes = set()
        for node, neighbours in self._adjacency_list.items():
            nodes.add(node)
            for neighbour in neighbours:
                nodes.add(neighbour)
        return frozenset(nodes)

    def neighbours(self, node):
        return self._adjacency_list.get(node, [])

    def in_degree(self, node):
        return self._in_degrees.get(node, 0)

    def out_degree(self, node):
        return self._out_degrees.get(node, 0)

    def is_node_one_in_one_out(self, node):
        return self.in_degree(node) == 1 and self.out_degree(node) == 1

    def _calc_node_degrees(self):
        in_degrees = defaultdict(int)
        out_degrees = defaultdict(int)
        for node, neighbours in self._adjacency_list.items():
            for neighbour in neighbours:
                out_degrees[node] += 1
                in_degrees[neighbour] += 1
        return in_degrees, out_degrees


def build_de_bruijn_graph(k_mers):
    d = defaultdict(list)
    for k_mer in k_mers:
        prefix = k_mer[:-1]
        suffix = k_mer[1:]
        d[prefix].append(suffix)
    return dict(d)


def build_nonbranching_paths(graph, node):
    paths = []
    for neighbour in graph.neighbours(node):
        w = neighbour
        path = [node, w]
        while graph.is_node_one_in_one_out(w):
            w = graph.neighbours(w)[0]
            path.append(w)
        paths.append(tuple(path))
    return paths


def find_nodes_with_outgoing_edge(graph):
    return {node for node in graph.nodes if graph.out_degree(node) > 0}


def build_cycle(graph, start_node):
    cycle = [start_node]
    node = start_node
    while (node := graph.neighbours(node)[0]) != start_node:
        cycle.append(node)
    cycle.append(start_node)
    return cycle


def find_isolated_cycles_in_graph(graph, found_paths):
    used_nodes = set(it.chain.from_iterable(found_paths))
    missing_nodes = set.difference(find_nodes_with_outgoing_edge(graph), used_nodes)
    isolated_cycles = []
    while missing_nodes:
        node = missing_nodes.pop()
        cycle = build_cycle(graph, node)
        isolated_cycles.append(tuple(cycle))
        missing_nodes.difference_update(cycle)
    return isolated_cycles


def find_maximal_non_branching_paths(graph):
    paths = []
    for node in graph.nodes:
        if (not graph.is_node_one_in_one_out(node)) and graph.out_degree(node) > 0:
            paths_from_node = build_nonbranching_paths(graph, node)
            paths.extend(paths_from_node)
    cycles = find_isolated_cycles_in_graph(graph, paths)
    paths.extend(cycles)
    return paths


def calc_string_spelled_by_a_genome_path(k_mers):
    text = list(k_mers[0])
    for k_mer in k_mers[1:]:
        text.append(k_mer[-1])
    return ''.join(text)


def main():
    k_mers = sys.stdin.read().splitlines()
    graph = Graph(k_mers)
    result = find_maximal_non_branching_paths(graph)
    print(' '.join(map(calc_string_spelled_by_a_genome_path, result)))


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter03/BA3K.scala
package TextbookTrack.Chapter03

object BA3K {
  import scala.annotation.tailrec

  class DeBruijnGraph[T](adjacencyList: Map[T, List[T]]) {
    private val (inDegrees, outDegrees): (Map[T, Int], Map[T, Int]) = calcNodeDegrees()

    lazy val nodes: Set[T] = adjacencyList.keySet ++ adjacencyList.valuesIterator.flatMap(_.toSet).toSet

    private def calcNodeDegrees(): (Map[T, Int], Map[T, Int]) = {
      val outDegrees: Map[T, Int] = adjacencyList.map{ case (node, neighbours) => (node, neighbours.length)}
      val inDegrees: Map[T, Int] =
        adjacencyList
          .iterator
          .flatMap{ case (node, neighbours) => neighbours.map(neighbour => (neighbour, node))}
          .toList
          .groupBy{ case (neighbour, _) => neighbour }
          .view
          .mapValues(_.length)
          .toMap
      (inDegrees, outDegrees)
    }

    def neighbours(node: T): List[T] = adjacencyList.getOrElse(node, Nil)

    def inDegree(node: T): Int = inDegrees.getOrElse(node, 0)

    def outDegree(node: T): Int = outDegrees.getOrElse(node, 0)

    def isNodeOneInOneOut(node: T): Boolean = (inDegree(node) == 1) && (outDegree(node) == 1)

    def nodesWithOutgoingEdges: Set[T] = nodes.filter(outDegree(_) > 0)
  }

  object DeBruijnGraph {
    def apply(kMers: List[String]): DeBruijnGraph[String] = {
      val adjacencyList: Map[String, List[String]] =
        kMers
          .map(kMer => (kMer.dropRight(1), kMer.drop(1)))
          .groupBy{ case (prefix, _) => prefix }
          .view
          .mapValues(_.map{ case (_, suffix) => suffix })
          .toMap
      new DeBruijnGraph[String](adjacencyList)
    }
  }

  private def buildNonBranchingPaths[T](graph: DeBruijnGraph[T], node: T): List[List[T]] = {
    def extendPath(node: T, neighbour: T): List[T] = {
      @tailrec
      def loop(path: List[T], w: T): List[T] = {
        if (graph.isNodeOneInOneOut(w)) {
          graph.neighbours(w) match {
            case List(x) => loop(x :: path, x)
            case _ => throw new Exception("Node should have only one neighbour.")
          }
        } else path.reverse
      }
      loop(List(neighbour, node), neighbour)
    }

    graph.neighbours(node).map(neighbour => extendPath(node, neighbour))
  }

  private def buildCycle[T](graph: DeBruijnGraph[T], startNode: T): List[T] = {
    @tailrec
    def loop(cycle: List[T], node: T): List[T] = {
      graph.neighbours(node) match {
        case List(nextNode) =>
          if (nextNode != startNode) loop(nextNode :: cycle, nextNode)
          else (startNode :: cycle).reverse
        case _ => throw new Exception("Node should have only one neighbour.")
      }
    }
    loop(List(startNode), startNode)
  }

  private def findIsolatedCyclesInGraph[T](graph: DeBruijnGraph[T], paths: List[List[T]]): Set[List[T]] = {
    @tailrec
    def loop(isolatedCycles: Set[List[T]], unusedNodes: Set[T]): Set[List[T]] = {
      if (unusedNodes.isEmpty) isolatedCycles
      else {
        val node: T = unusedNodes.head
        val cycle: List[T] = buildCycle(graph, node)
        loop(isolatedCycles + cycle, unusedNodes -- cycle)
      }
    }
    val missingNodes: Set[T] = graph.nodesWithOutgoingEdges.diff(paths.flatten.toSet)
    loop(Set.empty[List[T]], missingNodes)
  }

  def findMaximalNonBranchingPaths[T](graph: DeBruijnGraph[T]): List[List[T]] = {
    val paths: List[List[T]] =
      graph.nodes
        .iterator
        .filter(node => !graph.isNodeOneInOneOut(node) && graph.outDegree(node) > 0)
        .flatMap(buildNonBranchingPaths(graph, _))
        .toList
    val cycles: Set[List[T]] = findIsolatedCyclesInGraph(graph, paths)
    paths ++ cycles
  }

  def calcStringSpelledByAGenomePath(kMers: List[String]): String = kMers match {
    case Nil => ""
    case x :: xs => (x.toList ::: xs.map(_.last)).mkString
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val kMers: List[String] = reader.toList
    val graph = DeBruijnGraph(kMers)
    val result: List[List[String]] = findMaximalNonBranchingPaths(graph)
    println(result.map(calcStringSpelledByAGenomePath).mkString(" "))
  }
}
